### PR TITLE
Make direct item selection keys freely bindable

### LINKED ISSUE #4536: Let players freely bind keys to any action
The issue here is that you cannot rebind certain actions to other keys. The actions in question are:
- [ ] Attack/mine (Left mouse button)
- [ ] Build/use (Right mouse button)
- [x] Select next item in hotbar (Rotate mouse wheel down)
- [x] Select previous item in hotbar (Rotate mouse wheel up)
- [X] Select item X in hotbar (0-9)
- [ ] Open pause menu (Escape)

I request that Minetest allows to rebind keys freely for these actions.
For “Select item X in hotbar”, Minetest could make available 23 key bindings (e.g. `keymap_hotbar_1`, `keymap_hotbar_2`, etc.) because the hotbar can have 23 slots at maximum. Slots 1-9 are bound to the keys 1-9, and slot 10 is bound to key 0 by default, like now. The slots 11-23 are unbound by default. This allows interested players for more customization if the default 10 bindings aren't enough for them. :-)
I consider rebinding of Escape to be of minor importance. If it is too difficult to make this key freely re-bindable, it is okay to skip this.

See also: #2361 (binding mouse buttons to actions)

### PR DESCRIPTION
**Pull request by Wuzzy2**
This PR adds the new keymap settings `keymap_slot1`, `keymap_slot2`, etc. until `keymap_slot23`.
These settings allow the player to change the direct access keys to the hotbar slots for all possible slots, using `minetest.conf`.

The default key bindings for slots 1-10 are preserved. That is, press 1 for slot 1, 2 for slot 2, etc. and 0 for slot 10. 
Therefore, this will only be a change to players who actively edit `minetest.conf`.

Partially fixes #4536.

### CODE REVIEW

**Comment by SmallJoker** [comment]
Sorry but this looks way too overengineered to me. For me it would look better to make a simple boolean setting to turn the assignements of these numeric keys off.

**Comment by Wuzzy2** [comment]
@SmallJoker: I dislike your suggestion. We should allow players to rebind all keys no matter what. It makes Minetest less accessible to force the player to use certain configurations, or to limit the configuration of keys arbitrarily. A boolean setting is such an artificial limitation.

I will, however, accept remarks about the way I coded this (style, etc.).

Proof why this is needed:

> It may seem odd, but I would surely rebind 1-8 to tcv5rgwa. I always use the same keys in different games, except Minetest because of this.

(said ghost)

File: src/game.cpp [hunk 1]
```diff
@@ -1455,6 +1455,30 @@ void KeyCache::populate()
...
+	key[KeyType::SLOT_14]   = getKeySetting("keymap_slot14");
+	key[KeyType::SLOT_15]   = getKeySetting("keymap_slot15");
+	key[KeyType::SLOT_16]   = getKeySetting("keymap_slot16");
+	key[KeyType::SLOT_17]   = getKeySetting("keymap_slot17");
+	key[KeyType::SLOT_18]   = getKeySetting("keymap_slot18");
+	key[KeyType::SLOT_19]   = getKeySetting("keymap_slot19");
+	key[KeyType::SLOT_20]   = getKeySetting("keymap_slot20");
+	key[KeyType::SLOT_21]   = getKeySetting("keymap_slot21");
+	key[KeyType::SLOT_22]   = getKeySetting("keymap_slot22");
+	key[KeyType::SLOT_23]   = getKeySetting("keymap_slot23");
```
Comment (sfan5) [inline src/game.cpp]: `for()` loop

File: src/game.cpp [hunk 2]
```diff
@@ -2968,16 +2989,10 @@ void Game::processItemSelection(u16 *new_playeritem)
 		*new_playeritem = *new_playeritem > 0 ? *new_playeritem - 1 : max_item;
 	// else dir == 0
 
-	/* Item selection using keyboard
+	/* Item selection using hotbar slot keys (default: number keys 0-9)
```
Comment (sfan5) [inline src/game.cpp]: the default value doesn't need to be in this comment

**Comment by Wuzzy2** [comment]
@sfan5: Done.

File: src/game.cpp (hunk 1 above)
Comment (Wuzzy2) [inline src/game.cpp]: Yeah, I expected a comment like this. :D

File: src/game.cpp [hunk 3]
```diff
@@ -1455,6 +1455,11 @@ void KeyCache::populate()
 
 	key[KeyType::DEBUG_STACKS]   = getKeySetting("keymap_print_debug_stacks");
 
+	for(int i=0; i<23; i++) {
```
Comment (kaeza) [inline src/game.cpp]: `<24` *

**Comment by Calinou** [comment]
It's pretty common for competitive games to have this, so I'll :+1: it, even if Minetest is not a very competitive game.

File: src/game.cpp [hunk 4]
```diff
@@ -1455,29 +1455,10 @@ void KeyCache::populate()
...
-	key[KeyType::SLOT_15]   = getKeySetting("keymap_slot15");
-	key[KeyType::SLOT_16]   = getKeySetting("keymap_slot16");
-	key[KeyType::SLOT_17]   = getKeySetting("keymap_slot17");
-	key[KeyType::SLOT_18]   = getKeySetting("keymap_slot18");
-	key[KeyType::SLOT_19]   = getKeySetting("keymap_slot19");
-	key[KeyType::SLOT_20]   = getKeySetting("keymap_slot20");
-	key[KeyType::SLOT_21]   = getKeySetting("keymap_slot21");
-	key[KeyType::SLOT_22]   = getKeySetting("keymap_slot22");
-	key[KeyType::SLOT_23]   = getKeySetting("keymap_slot23");
+	for(int i=0; i<23; i++) {
```
Comment (sfan5) [inline src/game.cpp]: `for (int i = 0; i < 23; i++) {`

**Comment by SmallJoker** [comment]
@Wuzzy2 
> It may seem odd, but I would surely rebind 1-8 to tcv5rgwa. I always use the same keys in different games, except Minetest because of this.

That's exactly the reason why I thought this is was overengineered. I did not expect that people would want to mix the logical key numbering sequence. It's something I would never do, just because it doesn't fit with the numbering of the slots.
But as there are other people who would like to have this feature: :+1:

File: src/game.cpp (hunk 4 above)
Comment (Wuzzy2) [inline src/game.cpp]: lol.
Done.

**Comment by tobyplowy** [comment]
@Wuzzy2 +1

**Comment by rubenwardy** [comment]
this makes it easier for me to implement #5348, as it's a pain to mix Irrlicht and Action codes atm

(I had to do this)

**Comment by paramat** [comment]
:-1: Just put the items in the hotbar and use the existing keys.

**Comment by Wuzzy2** [comment]
So you are suggesting to add even more hardcoded keys? No, please don't. Making an entire row of letters (!) hardcoded is a recipe for desaster. Or did I misunderstand you?

> At most there are only 10 sequential keys free on a keyboard row (QWERTYUIOP)

This argument falls apart when you realize there are other keyboard layouts apart from US. But I do not actually want to argue *which* keys to use.

I think you need to see the bigger picture here: The big goal behind PRs like this is accessiblity. One way to achieve this is to eventually get away from all hardcoded keys in Minetest because hardcoded keys are never good. There are way too many keyboard layouts around and it is impossible to find a default setting which suits all needs. This is the very reason why we have configurable keys in the first place.

This PR does not actually introduce any new keys, it just ALLOWS the user to configure keys for slots 11-23 as well. I also don't intent to introduce new defaults for slots 11-23 (for now) because it would lead to a weird default. But users would be free to use this if they want, with this PR.

Intentionally leaving out 3 keys because of the number 23 is a very strange idea. No harm whatsoever is done for adding 3 more *bindable* keys.

> > (…)

> That's not a good argument

Of course it is! I posted this to refute the claim that no one ever would need this.

**Comment by Zeno-** [comment]
Not everyone's keyboard is QWERTYUIOP :/

**Comment by sofar** [comment]
:+1: patch seems entirely reasonable, code looks sane too

This is a small step to making input more generic and flexible, so I don't understand the big fuss about this.

Heck, kudos for going to the max instead of stopping at 10.

**Comment by paramat** [comment]
> add 10 extra keys for slots 11-20

Sorry i mean 10 extra configurable keys.
I can see the argument for making all slot keys configurable, it's just i can't see any other keys than 1-0 making sense for slots 1-10.
By QWERTYUIOP i meant the top row of letter keys on a keyboard, whatever that is, there's 10 letter keys whereas there's only 9 on the row below.
Maybe a -1 is too extreme here, i'll remove it (and document that i removed it to avoid upsetting kilbith).

**Comment by paramat** [comment]
kilbith,

I did indeed encourage SmallJoker to stick with his opinion instead of letting other's opinions override his own, it was a type of encouragement. My first version of this comment was written while i was in a particularly bad early-morning mood, a little later i felt a little better and realised it was perhaps unreasonable to encourage another dev to -1 the PR so i removed it, and edited other unreasonable parts of the comment.

I didn't want unreasonable and negative comments to persist so i deleted them. I expected the comments had already been seen by many so i had no hope of 'getting away with it'. I am not going to start leaving everything i write in place because it is a waste of space and means unreasonable, negative or unpleasant comments remain to be read, it's best to delete these negative things. If i no longer stand behind a comment then that comment is obviously best deleted.

It is done because i believe in removing negativity, not because i hope no-one saw it and i can get away with it, as you are stating on IRC, i rarely 'get away with it' as notifications contain the original and many are quick to read comments.
It's not good to try to rouse a lynch mob on IRC.

**Comment by VanessaE** [comment]
This is all offtopic for this particular issue, but if you have to make drastic edits to remove content you no longer stand behind, use ~~strikethrough~~ instead of rewriting the post.

**Comment by Ekdohibs** [comment]
:+1: for this PR, especially since not everyone uses a qwerty layout.

**Comment by paramat** [comment]
Well, strikethrough leaves it readable so i wouldn't be happy with doing that.

**Comment by VanessaE** [comment]
Readable though it may be, it also clearly tells the reader that the text in question is no longer applicable.  Meanwhile, :+1: to this PR.  I have had players asking me if there were keys to access the whole 16-slot Dreambuilder hotbar as well.

**Comment by paramat** [comment]
I'm neutral on the PR.

**Comment by sofar** [comment]
> it's just i can't see any other keys than 1-0 making sense for slots 1-10.

I can think of various reasons:

- several subgames (carbone I think? I've seen others) have double the hotbar slots
- I could see a subgame designate certain slots as special and reserving them for special tools (portal gun, off hand weapon, shield etc.) and you'd use specially bound keys to reach them directly without affecting contents of the main hand.
- there's 14 keys at the top of my keyboard, not 10. 2 are likely unusable, that leaves 12 keys. Logically, a hotbar with 12 items seems entirely reasonable.

**Comment by sofar** [comment]
I'm counting 2 approvals but it needs rebase. perhaps @sfan5 can merge to indicate the requested changes were made and he's ok with it.

**Comment by Wuzzy2** [comment]
Wait, this needs ANOTHER rebase? I just rebased it yesterday. :-(

**Comment by est31** [comment]
@paramat I'd have loved to talk about this on IRC, but generally you should try to think twice before you make a comment, or maybe wait with it until you are in a better mood. I know there are so many things being suggested all the time, and many of them have issues. But it would be really a bit better for everyone if you could edit your posts less often. I admit to have done it sometimes myself as well, but I always feel a bit guilty when I do it, as its not good style.


About the proposed change itself. I think the limit 23 is a bit arbitrary, it could be 11 or 33 just as well. I don't have any better idea for a system that doesn't impose such an arbitrary limit though, other than suggesting that only one setting should be used which then contains a ; separated list of key bindings. But that change would have its own issues. It would be a bit user unfriendly as you can't individually set keys to the default, and not fit into our current system of how key bindings are assigned. Probably not everything needs to be perfect, idk.

**Comment by Wuzzy2** [comment]
23 was chosen because this is the maximum supported hotbar size. If the maximum would be 20, I would have picked 20. Simple as that.

**Comment by Wuzzy2** [comment]
Are there any reasons which prevent this from being merged?

I don't know how to “rebase” a single commit only. What do you want?

**Comment by paramat** [comment]
'Rebase needed' doesn't mean 'squash', it means changes have been made in MT which conflict with your changes, so you need to 'base-your-code-on-master'.

**Comment by Wuzzy2** [comment]
I hope this is better now.

**Comment by paramat** [comment]
640ba7727629392211ed3c2bd116e938883a6f90
@Wuzzy2 docs need adding to minetest.conf.example, you can auto-generate it though.

**Comment by Wuzzy2** [comment]
WTF? I could swear I have put this into `minetest.conf.example`. I'll check tomorrow.

**Comment by Wuzzy2** [comment]
You can grab the lastest commit now which fixes `minetest.conf.example`.

I don't know why I didn't update it earlier, maybe I screwed up a merge or something.

**Comment by paramat** [comment]
2nd commit is good :+1:

**Comment by SmallJoker** [comment]
Doc fix: 86c0b27

**Pull request closed without merging**